File: testdata/taint_sample.py
import subprocess
from flask import request, render_template

def taint_direct():
    """User input flows directly to SQL."""
    query = request.args.get("q")
    db.execute(query)

def taint_via_call():
    """User input flows through helper to subprocess."""
    cmd = request.args.get("cmd")
    run_command(cmd)

def run_command(cmd):
    subprocess.run(cmd, shell=True)

def taint_template():
    """User input flows to template render."""
    name = request.args.get("name")
    return render_template("hello.html", name=name)

def no_taint(x, y):
    """No user input, no taint."""
    return x + y

File: testdata/taint_sample.go
package testdata

import (
	"database/sql"
	"net/http"
	"os/exec"
)

// TaintDirect: user input flows directly to SQL sink.
func TaintDirect(w http.ResponseWriter, r *http.Request) {
	query := r.URL.Query().Get("q")
	db, _ := sql.Open("postgres", "")
	db.Query(query)
}

// TaintViaCall: user input flows through a helper to SQL sink.
func TaintViaCall(w http.ResponseWriter, r *http.Request) {
	input := r.URL.Query().Get("cmd")
	runQuery(input)
}

func runQuery(q string) {
	db, _ := sql.Open("postgres", "")
	db.Exec(q)
}

// TaintUnreachable: user input with length check.
func TaintUnreachable(w http.ResponseWriter, r *http.Request) {
	input := r.URL.Query().Get("x")
	if len(input) > 100 {
		return
	}
	_ = input
}

// NoTaint: no user input, no taint path.
func NoTaint(x int, y int) int {
	result := x + y
	return result
}

// TaintToCommand: user input flows to command execution.
func TaintToCommand(w http.ResponseWriter, r *http.Request) {
	cmd := r.URL.Query().Get("cmd")
	exec.Command(cmd).Run()
}
